src/aliceVision/matching/ArrayMatcher_hnswlib.hpp

```diff
@@ -0,0 +1,167 @@
+// This file is part of the AliceVision project.
+// Copyright (c) 2016 AliceVision contributors.
+// Copyright (c) 2012 openMVG contributors.
+// This Source Code Form is subject to the terms of the Mozilla Public License,
+// v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "aliceVision/matching/ArrayMatcher.hpp"
+#include "aliceVision/matching/CascadeHasher.hpp"
+#include "aliceVision/matching/IndMatch.hpp"
+
+#include <aliceVision/system/Logger.hpp>
+
+#include "dependencies/hnswlib/hnswlib/hnswlib.h"
+
+#include <memory>
+#include <random>
+#include <cmath>
+
+using hnswlib::HierarchicalNSW;
+using hnswlib::L2Space;
+using hnswlib::L2SpaceI;
+using hnswlib::SpaceInterface;
+using namespace std;
+
+namespace aliceVision
+{
+namespace matching
+{
+
+//------------------
+//-- Bibliography --
+//------------------
+//- [1] "Fast and Accurate Image Matching with Cascade Hashing for 3D Reconstruction"
+//- Authors: Jian Cheng, Cong Leng, Jiaxiang Wu, Hainan Cui, Hanqing Lu.
+//- Date: 2014.
+//- Conference: CVPR.
+//
+
+// Implementation of descriptor matching using the cascade hashing method of [1].
+// If you use this matcher, please cite the paper.
+// template Metric parameter is ignored (by default compute square(L2 distance)).
+template <typename Scalar = float, typename Metric = L2_Simple<Scalar>>
+class ArrayMatcher_hnswlib : public ArrayMatcher<Scalar, Metric>
+{
+public:
+    typedef typename Metric::ResultType DistanceType;
+
+    // Some initialization
+    ArrayMatcher_hnswlib() {}
+    virtual ~ArrayMatcher_hnswlib() {}
+
+    /**
+     * Build the matching structure
+     *
+     * \param[in] dataset   Input data.
+     * \param[in] nbRows    The number of component.
+     * \param[in] dimension Length of the data contained in the dataset.
+     *
+     * \return True if success.
+     */
+    bool Build(const Scalar* dataset, int nbRows, int dimension)
+    {
+        HNSWmetric.reset(nullptr);
+        HNSWmatcher.reset(nullptr);
+        dimension_ = 0;
+
+        if(nbRows < 1)
+            return false;
+
+        dimension_ = dimension;
+
+        // Here this is tricky since there is no specialization
+        if(typeid(DistanceType) == typeid(int))
+        {
+            HNSWmetric.reset(dynamic_cast<SpaceInterface<DistanceType>*>(new L2SpaceI(dimension)));
+        }
+        else if(typeid(DistanceType) == typeid(float))
+        {
+            HNSWmetric.reset(dynamic_cast<SpaceInterface<DistanceType>*>(new L2Space(dimension)));
+        }
+        else
+        {
+            std::cerr << "HNSW matcher: this type of distance is not handled Yet" << std::endl;
+            return false;
+        }
+
+        HNSWmatcher.reset(new HierarchicalNSW<DistanceType>(HNSWmetric.get(), nbRows, 16, 100));
+        HNSWmatcher->setEf(16);
+
+        // add first point..
```

Comment:
Why did you trim OpenMP? I think you'll experiencing performance loss.

Reply:
Thank you for pointing this out. We did not use it because we followed the same format for the `#include` statements and general file layout as the other `ArrayMatcher` files in Alice Vision for matching. We will add OpenMP, test it, and then push the changes for you to see here.